01-messaging.md

```diff
@@ -476,6 +476,77 @@ every message maximally).
 
+### Peer storage
+
+## The `peer_storage` and `your_peer_storage` Messages
+
+Nodes that advertise the `option_provide_storage` feature offer storing
+arbitrary data for their peers. The data stored must not exceed 65531 bytes,
+which lets it fit in lightning messages.
+
+Nodes that advertise the `option_want_storage` feature want to have their
+data stored by their peers that support `option_provide_storage`.
+
+Nodes can verify that their `option_provide_storage` peers correctly store
+their data at each reconnection, by comparing the contents of the
+retrieved data with the last one they sent. However, nodes should not expect
+their peers to always have their latest data available.
+
+Nodes ask their peers to store data using the `peer_storage` message & expect peers to return them latest data using `your_peer_storage` message:
+
+1. type: 7 (`peer_storage`)
+2. data:
+   * [`u16`: `length`]
+   * [`length*byte`:`blob`]
+
+
+1. type: 9 (`your_peer_storage`)
+2. data:
+   * [`u16`: `length`]
+   * [`length*byte`:`blob`]
+
+
+Requirements:
+
+The sender of `peer_storage`:
+  - MUST support `option_want_storage`.
+  - MAY send `peer_storage` to only peers who support `option_provide_storage`.
+  - MAY send `peer_storage` whenever necessary (usually, everytime their configuration changes)
+  - MUST limit its `blob` to 65531 bytes.
+  - MUST encrypt the data in a way that can be validated when received.
+
+
+The receiver of `peer_storage`:
+  - If it offered `option_provide_storage`:
+    - if it has an open channel with the sender:
+      - MUST store the message.
+    - MAY store anyway (e.g. might get paid for the service).
+    - MUST replace the old `blob` with the latest received.
+
+- If it does store the message:
+  - MUST immediately respond with `your_peer_storage` containing `blob`
+  - MUST send `your_peer_storage` again after reconnection.
+- Otherwise:
+  - MUST NOT respond with `your_peer_storage`
+
+The sender of `your_peer_storage`:
+  - MUST include the last `blob` it stored for that peer.
+  - when all channels with that peer are closed:
+    - SHOULD wait at least 2016 blocks before deleting the `blob`.
+
+The receiver of `your_peer_storage`:
+  - when it receives `your_peer_storage` with an outdated or irrelevant data:
+    - MAY send a warning.
+
+Rationale:
+If a peer supports `option_provide_storage` and we have a channel with it, we can send a `peer_storage` message with any `blob` so that peer can store it locally in their node. Everytime a node receives a `peer_storage` message, It needs to update the `blob` for the sender.
+
+Upon receiving `peer_storage` nodes should immediately respond with the latest `your_peer_storage`, which could work as an ack to the peer.
+
+Nodes with `option_provide_storage` should also regularly send `your_peer_storage` to verify that they store the latest and correct form of `blob`. Nodes may want to ignore `peer_storage` messages with peers they don’t have a channel with to avoid spam.
```

Comment:
This part is unclear to me. Based on my reading, there are two instances when a node (running `option_provide_storage`) will send `your_peer_storage` message:
- When it gets a `peer_storage` message and has stored the data (like an ack)
- When it reconnects with the peer for which it has stored data (to ensure that peer has a chance to update stale data)

The point about "regularly" sending `your_peer_storage` seem to be specified loosely.
- What's the right frequency to send this to your peer?
- Does it create a spam vector for your peer?

Reply:
Agreed. You don't need to specify this, since you already said in requirements that "MAY send `peer_storage` whenever necessary (usually, everytime their configuration changes)".

And I made a suggestion for spam above...

Reply:
Yes, the two instances that you mentioned are the only time when a peer needs to send `your_peer_storage`.